Initial work on 3D vision

Coolthulhu opened the pull request:
Mostly pushing so that I can get a design issue/question resolved:
Current `line_to` does diagonal moves. This is OK for x/y, but diagonal moves that include z-level moves make my job harder.
Is it OK to have `line_to` split horizontal (x/y) and vertical moves into one-then-other? Is my version (the one in the diff) of splitting it correct?

Now the actual summary of the PR:
- Makes `map::build_map_cache` build transparency and outside caches for all z-levels (in z-level mode). This is actually a minor bugfix, as it could have been enabling some subtle (minor) bugs.
- Makes the game not recalculate sight cache when looking around (in look mode) when it doesn't change (still recalculates a lot, there's room for improvement)

Adds an option in debug tab. Setting it on enables the stuff below:
- 3D `map::sees` and `map::find_clear_path`. Those work like 2D ones except they use 3D bresenham,  and when making a z-move they check for `m.valid_move` to determine if it can actually be made.
- Adds a slow raycasting 3D FoV calculation that is used only when the debug option is on. It checks all potentially visible tiles. Based on a stripped-down version of the proper shadowcasting algorithm. Megatons of room for improvement here, this is just for testing for now.

For now I'll be going for #6824 (sight and attacks) rather than #6821 (actual FoV itself), as the former is both more noticeable and easier to implement.

kevingranade commented:
What about diagonal+vertical offsets makes things harder?
line_to() should just draw a bresenham line to the target, if something needs some other constraints, it needs to be a separate function.  It was a real mess cleaning out the weirdness around that from the original.

kevingranade commented on src/lightmap.cpp:
```diff
@@ -490,98 +536,126 @@ void map::build_seen_cache( const tripoint &origin, const int target_z )
+            return !valid_move( p, below, false, true );
+        };
+
+        // Raycasting, as opposed to shadowcasting for 2D
```
Raycasting is immensely slower than shadowcasting (and it has weird artifacts), I'm extremely skeptical that this will be a workable solution for anything other than working on the draw code.  The 3d FoV code is going to need to be either shadowcasting, or something more sophisticated than shadowcasting.  Probably we can get by with shadowcasting plus some z-level aware fixups.

Coolthulhu replied:
Yeah, that's why I am going for the part 5 with it first - just for sight and attacks, not as the FoV function.

Updating shadowcasting to 3D would require some clever algorithm - either extending shadowcasting to 3D (which would probably be a totally new algorithm) or tolerating some simplifications and asymmetries during z moves.

kevingranade replied:
My bad, it was late and I missed that part of your summary.
I think extending shadowcasting to 3D is doable, I have a mental model for it in my head.
The shadowcasting algorithm works by finding spans of adjacent squares within an octant and processing them recursively.  It does this by iteratively considering each square and deciding whether to add it to the current span, or use it as the start of a new span.  As an optimization (and the source of the name) spans with no visibility are not processed.  Before the algorithm processes a new span, it recursively processes the previous span.
To extend the algorithm to 3D, the algorithm would need to find 3D spans (frustums) with uniform transparency, it can do this by iterating along a major and minor axis, when a change in transparency is found, the algorithm will need to split the existing frustum along the major axis (creating a frustum with depth along both the major and minor axis) and split the previous segment of the current iteration along the minor axis (creating a frustum with no width along the major axis, and 0 or more width along the minor axis), the algorithm would recurse both frustums, then proceed to the new span.  I can see it in my head, but it's really hard to explain :/

Coolthulhu replied:
I see what you mean.
I had a similar idea.

Well, I'm not sure if I get what minor and major axes are here, but otherwise I think I get it.
If minor and major axes are arbitrarily chosen (like now y vs x in `castLight`) rather than depending on some factor, then I'm pretty sure I get it.

Coolthulhu replied:
One final check if I really get it:
We want to split the currently processed projection/rectangle/etc. into 3 parts:
* Rectangle that is already processed
* Rectangle that has one line processed, rest unprocessed. The line ends with (before? after?) a point with change in transparency
* Rectangle that is unprocessed

Then extend a frustum (another recursive call) from each of those rectangles.

Right?

kevingranade replied:
Yea, sounds like you have it.
The reason I used major vs minor is it depends on the region being
processed, sometimes it'll be x/y, sometimes x/z, sometime y/z, possibly
with different orderings.

chaosvolt commented:
Oh wow, this gets more impressive by the day.

Coolthulhu commented:
> What about diagonal+vertical offsets makes things harder?

A horizontal move has a known vision check (check transparency of the destination tile), vertical move also has a known vision check (existence of the floor on the upper tile).
Horizontal+vertical needs to be split into horizontal+vertical, with the original order of those not being known (ie. in bresenham it could be horizontal+vertical, vertical+horizontal or even h+v+h, but it can't be easily discerned on the outside).

Coolthulhu commented:
> line_to() should just draw a bresenham line to the target, if something needs some other constraints, it needs to be a separate function.

I'll split the `line_to` and `bresenham` into their original and modified versions then.
This presents a bit of a maintenance problem, though: functions like monster attacks, vision, projectiles, any sort of movement etc. will have to use the modified versions, the regular "diagonal z" bresenham will not actually be used in many functions.

Coolthulhu commented:
OK, split the bresenham and line_to into the old one and the modified one.

Added an option to see critters below the player, but it requires either the `fov_3d` or `debug_mode` to be on. I can't easily add a test for clairvoyance or DEBUG_VISION here, as those take quite a bit of time and aren't cached (actually, clair is cached, just that this cache is not accessible in `Creature::sees` or `player::sees`).

Added drawing a symbol for "critter is below here". A green 'v' on cyan background. Only in fov_3d and debug modes for now.

kevingranade commented:
I'm still not clear on the need for a modified bresenham, I'll read the code and try to figure it out, but I figured I'd try and clarify in prose first.
one, both diagonal+vertical and on-the-same-level+vertical (i.e. x+y+z vs x+z or y+z) have been mentioned, are these separate issues or the same issue?
Setting aside diagonal vs orthogonal, is the problem with transitioning horizontally and vertically at the same time that the order-of-operations for the checks is poorly-defined?  I can see there being a problem there, if moving up, you need to check for a roof above the source tile, but if moving down you need to check for a roof above the destination tile.  That seems doable in the caller if they have a normally rendered bresenham line, but I might be missing something.

The underlying problem is that horizontally we're using a cell or voxel model, but we're not using that model vertically, things that aren't cells can block vertical movement.  Therefore it's not sufficient to check the identity of the target square.

Coolthulhu commented:
> one, both diagonal+vertical and on-the-same-level+vertical (i.e. x+y+z vs x+z or y+z) have been mentioned, are these separate issues or the same issue?

Just horizontal+vertical, whether the horizontal part itself is diagonal or not doesn't really change anything.

> is the problem with transitioning horizontally and vertically at the same time that the order-of-operations for the checks is poorly-defined

Mostly this.

>  I can see there being a problem there, if moving up, you need to check for a roof above the source tile, but if moving down you need to check for a roof above the destination tile. That seems doable in the caller if they have a normally rendered bresenham line, but I might be missing something.

Moving (walking) is a relatively easy case, as it can be assumed that the higher point has a floor.
Consider vision, a flying bullet or even a flying creature. Those may want to execute the diagonal move by going horizontally first.

Additionally, the horizontal move still has to be checked. This leads to an additional split that looks like this:
```
if( z_level_changed ) {
    tripoint midpoint( new_point.x, new_point.y, old_point.z );
    return valid_move( old_point, midpoint ) && valid_move( midpoint, new_point );
}
```
Which would further slow down processing of "diagonally vertical" lines.

Rivet-the-Zombie commented:
Awesome stuff, @Coolthulhu!

kevingranade commented:
I think I see, boiling it down to having floating floors above the player's head (between z-levels 0 and 1), how do you decide which squares you can interact with via vision or a thrown object.  If there is a ceiling immediately above you, does that mean you can't interact with the 8 squares horizontally adjacent to you but one square up?  Similarly, if there is a ceiling above each adjacent square, but not one directly above you, can you still interact with those squares?  Intuitively the second must be true, since it corresponds to a player standing in a pit and trying to interact with things adjacent to the pit.  the question is, do we also want the first to be true, I think so.
I still think we don't want that difference encoded in the line-drawing function, we want it in things that call it, like creature::sees(creature) and similar.  This roughly corresponds to the thing where sees() used to call line_to() with a bunch of different start points to scan for a viable path.

> Which would further slow down processing of "diagonally vertical" lines.

I don't think there's a good way around that, your alternative extends the length of the path, which means it processes more squares instead of slightly increasing the cost per-square.

I think if deciding if 0,0,0 (source) can see/attack 1,0,1 (destination), you need to do something like:
```
if( source.z != destination.z ) {
    // Whether we're going up or down, the floor between source and destination is at the higher of the two z levels.
    int floor_z = std::max(source.z, destination.z);
    // Check for a clear path headed from the source path through the square at source x,y and with the z of the destination (z-first) or through the square at destination x,y with the z of the source (z/y-first).
    if( ( !floor_at(destination.x, destination.y, floor_x) &&
          trans(destination.x, destination.y, source.z) ) ||
        ( !floor_at(source.x, source.y, floor_z) &&
          trans(source.x, source.y, destination.z) ) ) {
        path_clear = true;
    }
}
```

Coolthulhu commented:
Now I've got a design problem:
The vision algorithm needs a 3D array.
Transparency cache is a 2D array, there is no 3D transparency cache.

I'm passing an array of pointers to 2D arrays, but that's kinda ugly.

Any better way? Preferably (but not necessarily) not involving redesigning the entire cache or dropping the ability to use the same function to project vision and lights.

kevingranade commented:
If you redeclare the cache as a ```float transparency_cache[NUM_ZLEVELS][MAPSIZE*SEEX][MAPSIZE*SEEY];`` I think you you can pass a particular zlevel of it with ```&transparency_cache[z][0][0]``` for the 2D code, and pass the whole thing for the 3D code, sound good?  That means either updating all the 2D methods to extract that field, or making an alias for it for either 2D or 3D.

You might have to cast the resulting pointer to pass it as a sized array instead of as a raw pointer, haven't checked.

Coolthulhu commented:
The implementation seems to be going well, but I still got some bugs to fix.
I have some problems with how to properly do the recursion/split the processed area.

I iterate over y, then z, then x. This means that the area is processed in squares/rectangles up to d tiles tall and up to d tiles wide, where d is the square distance to player.
y and x aren't necessarily y and x in game, because the shadowcasting algorithm can be rotated. z coord is always z coord (for now?).

A given rectangle being processed looks like this:
```
CCCDDD
CCCDDD
BBBDDD
AAAAAA
AAAAAA
```
* A - the rectangle already processed. Here we can safely recurse with y distance increased by 1
* B - a line (thickness 1 rectangle) already processed that can't be a part of A because it isn't as wide.
* C - the rectangle not yet processed above B. Could possibly contain B?
* D - the rectangle not yet processed not above B.

When I encounter a transparency change, I recurse on A, B and C, then process D in current function.
A and B have distance (y coord displacement) increased by 1 compared to currently processed rectangle. C and D are clipped.

Anything I'm doing wrong? Should I document the algorithm somewhere?

kevingranade commented:
That looks right.
I assume in your example there was a transparency change between B and D.
At this point I'd suggest recursing on A and B, trying to merge B and C is
likely more trouble than it's worth.

kevingranade commented on src/lightmap.cpp:
```diff
@@ -469,6 +470,171 @@ bool map::pl_sees( const tripoint &t, const int max_range ) const
+        return;
+    }
+
+    /*
```
I don't think we ever want something like this, we always want to determine what the player can see, whether we draw it or not.  I might be misunderstanding though.

kevingranade commented:
BTW, take a peek at tests/shadowcasting_test.cpp, it's not working perfectly, but it has some basic infrastructure for a few things.

Coolthulhu commented:
> I don't think we ever want something like this, we always want to determine what the player can see, whether we draw it or not. I might be misunderstanding though.

This is mostly for speed now. The 3D casting is unoptimized and because of that unusable for full FoV scan. Later on when I get rid of the horrid overdraw (up to 32 times at +10 zlevels) and add a floor cache, I'll try without it.

> BTW, take a peek at tests/shadowcasting_test.cpp, it's not working perfectly, but it has some basic infrastructure for a few things.

Could be useful for performance testing. Not sure what to look for in the file, though.
I'll probably add a test to verify whether 3D casting produces results similar to 2D casting (up to an epsilon).

chaosvolt commented:
It looks like you also need to update the master and such, its puking up "previous definition is here" errors due to https://github.com/CleverRaven/Cataclysm-DDA/commit/aeb55a0517ebcd8810b27e0449e03971087a4ac1 apparently.

Coolthulhu commented:
Floors working with the following limitations:
* You can see the first blocking tile. That is, when you look at a sewer, you can see stuff there.
* There is some weird bug that excludes single lines of vision when looking outside current z-level. Doesn't happen a lot. Must be something with recursion.
* Sloooow. Looking 5+ z-levels up is asking for trouble.
* Vision is a bit jagged when looking at non-uniform terrain. For example, looking at high buildings with lots of glass windows will cause jaggies to appear in vision above the building.

I inverted one assumption in the code: `delta.z` iterates from `0` to `distance`, unlike `delta.x` which goes from `-distance` to `0`.
Is there a particular reason why x iteration was inverted? This inverted iteration seemed to cause some problems for z-coord iteration.

Is the iteration un-inversion OK?

chaosvolt commented:
Oh right, I forgot I made that earlier remark before discovering the cause of the earlier shenanigans.

Coolthulhu commented:
The vision itself should be up for testing. I'd like to make sure it's acceptable before moving to easier things (extending functionality, optimization etc.).
To enable, open options->debug and set the last option to true.

What is working:
* Vision itself
* Floors obscuring vision
* Vehicle roofs working as floors

What isn't working yet:
* Floors NOT obscuring vision when it would make sense for them not to obscure it (for example, when viewing tile 10 tiles away and just 1 z-level above us)
* Performance. It dies when looking up (becomes unacceptable around +5 z-levels)
* Multiple z-level cache output. This means no 3D lights yet.
* Seeing monsters. It depends on seen cache, which is currently written one z-level at a time for ease of testing (having tolerable performance), meaning it currently works through floors and depends on last viewed z-level
* Looking directly up and directly down
* Transparent floors

kevingranade commented:
A possible optimization, perhaps we can cache the floor and ceiling of what is potentially visible, this will mostly optimize for the common case of walking around on the surface, where there is nowhere with vision to a lower z-level.  Meanwhile, whenever we break this assumption, we generally have a LOS-breaking ceiling above us, and also restricted LOS in general since we're down in a structure.  Any amount of underground shadowcasting *should* be trivial since it prunes all but a very small volume around the player.
This might, at least in the very common case prevent the very expensive double shadowcast of z = u.posz(), although again that gets thrown away if the player is at all elevated.

kevingranade commented:
Crazy thought while testing, since the top of the un-scanned area directly above the player is only 17x17 (the whole volume is 1491 squares, vs 14641 for *just* 2D shadwcasting on z-level 0), do we really need to handle that with shadowcasting?  Maybe we can use a slightly less optimal algorithm that handles the common case of scanning that pyramid of open air very quickly.  Similarly the "check below your feet"  check is almost always going to dead-end immediately, all you have to check for is that the 3x3 area directly below the player is opaque.
Another crazy thought for avoiding neagitive-z shadowcasting, possibly the transparency cache can check for floor transparency while it's doing its thing, and if every floor square on the player's square in a given octant is opaque, it skips negative-z sladowcasting in that octant.  If it helps, we could do this check for every level and possibly do some shortcuting.

Very minor issue I noticed testing, when you look around and change levels, it draws the level above you, then draws it centered on your aim point.  That's likely already happening without the 3D changes, but fast enough that it's not visible.

kevingranade commented on src/game.cpp:
```diff
@@ -4982,6 +4982,16 @@ void game::draw_critter( const Creature &critter, const tripoint &center )
+            // TODO: Make this show something more informative than just green 'v'
+            // TODO: Allow looking at this mon with look command
+            // TODO: Redraw this after weather etc. animations
+            mvwputch( w_terrain, my, mx, c_green_cyan, 'v' );
```
Yipes, the green 'v' is almost invisible on my monitor, possibly a different color for the background?  Alternately a dark color for the 'v' would contrast better.

kevingranade commented on src/lightmap.cpp:
```diff
@@ -469,6 +470,173 @@ bool map::pl_sees( const tripoint &t, const int max_range ) const
+    float (&output_cache)[MAPSIZE*SEEX][MAPSIZE*SEEY],
+    const std::array<const float (*)[MAPSIZE*SEEX][MAPSIZE*SEEY], OVERMAP_LAYERS> &input_arrays,
+    const tripoint &offset, const int offset_distance, const int target_z,
+    const std::function<bool(const tripoint &)> &floor_check,
```
Big red flag here (floor_check), the reason I supplied calc and check as template arguments is that dispatching to std::function objects is very expensive in a loop this tight.  I had no idea it was as bad as it is before I tested it, but it's BAAD.

chaosvolt commented:
I was starting to worry that the details of getting 3D vision to work would compound things and make it a more hefty process, yeah...

Coolthulhu commented:
> A possible optimization, perhaps we can cache the floor and ceiling of what is potentially visible, this will mostly optimize for the common case of walking around on the surface, where there is nowhere with vision to a lower z-level.

Yeah, I do intend to add something like this. A floor/ceiling cache similar to outside cache, covering the entire bubble.
The cache could have a downscaled (ORed) version, which then would be used to determine if casting up/down is actually needed and in which direction.

> Big red flag here (floor_check), the reason I supplied calc and check as template arguments is that dispatching to std::function objects is very expensive in a loop this tight.

The reason floor_check is a function object right now is that I wanted to defer optimization until the shape of the FoV is acceptable.

kevingranade commented:
It *is* exponentially more work, but there are opportunities to be more clever about it, so I think we can manage.  There are several more levels of optimization we haven't touched yet.

kevingranade commented:
Oh, random bug, I was tootling around on top of the shelter, there was a mi-go and some krecks on the ground nearby.  When I jumped off the roof of the shelter I got messages saying the mi-go and krecks fell.

Coolthulhu commented:
Yeah, I'm pretty sure even just manipulating the order of variables can help a lot. While I was trying to fix one of my bugs, I inverted the order of z-coord iteration. This caused 3x increase in execution time due to recursion having to split the frustum into thin, tall, long frustums and then those into lines (as opposed to one-two big blocks and then some long lines).

>  When I jumped off the roof of the shelter I got messages saying the mi-go and krecks fell.

I couldn't replicate in a simple test. May be related to monster spawning just before a vertical shift or something like that.

Coolthulhu commented:
In case anyone is wondering why is it going so slowly, I'm having problems with what looks like the last edge case:

![The FoV edge case](http://s11.postimg.org/5luy89usz/dda_fov3d.png)

What I figured out:
* It only happens when looking up/down at least 2 z-levels (image is how it looks when looking from a 2 z-level tall building)
* It doesn't happen to some near edges, only to ones further away. Those edges which are drawn properly share the common trait: one of their ends can be reached by going only in diagonals from player's tile. Suggests recursion being wrong?
* Didn't manage to fix it by fiddling with slope precision

This is the worst kind of programming - "shotgun programming". I'm at the point where I'm changing (semi)random stuff and checking that it still doesn't work.

chaosvolt commented:
Oh wow. All of this stumping me is expected, but this takes the cake.

kevingranade commented:
I caused a bug resembling this many times when fiddling with the shadowcasting code, it was always due to me accidentally resetting newStart.

BevapDin commented on src/lightmap.cpp:
```diff
@@ -490,98 +673,160 @@ void map::build_seen_cache( const tripoint &origin, const int target_z )
 
-            const auto mirror_pos = veh->global_pos() + veh->parts[mirror].precalc[0];
+        // Down
+        cast_zlight<0, 1, 0, 1, 0, 0, -1, sight_calc, sight_check>(
```
All calls to `cast_zlight` have the same value for the template parameter `xz` and `yz`: 0. Is that intended? Is there actually any need for those parameters (extensions planed for the future)? Also, the // Up and // Down comments are strange, why are there 10 calls to `cast_zlight` after the // Down, but only 6 calls to it below the // Up comment?

Coolthulhu replied:
> All calls to cast_zlight have the same value for the template parameter xz and yz: 0. Is that intended? Is there actually any need for those parameters (extensions planed for the future)?

I added them before I knew what I needed, so they may actually be useless. I thought I might be able to use them to map the small cones directly above and directly below the player, but it doesn't really sound like it would work.

> Also, the // Up and // Down comments are strange

I got the comments wrong there, what matters is the `zz` template parameter. `// Up` should be above the first `cast_zlight` with `zz == 1`.

Coolthulhu commented:
While this isn't done, it performs tolerably for most cases.
I think it can be tested and merged. I'll come back to finish it later, after #6824, which is easier than fixing all the edge cases in FoV (in fact, I already did most of #6824 while implementing other stuff).

Current issues are:
* "Jaggies" when looking up or down near a wall. Something with new slopes not being consistent
* Slow
* When looking up, all terrain with floors is obscured. You couldn't see a hulk looking at you from the roof.

stk2008 commented:
Ohhhh cant wait to test this one :)

On Sat, Jan 9, 2016 at 11:08 AM, Coolthulhu <notifications@github.com>
wrote:

> While this isn't done, it performs tolerably for most cases.
> I think it can be tested and merged. I'll come back to finish it later,
> after #6824 <https://github.com/CleverRaven/Cataclysm-DDA/issues/6824>,
> which is easier than fixing all the edge cases in FoV (in fact, I already
> did most of #6824
> <https://github.com/CleverRaven/Cataclysm-DDA/issues/6824> while
> implementing other stuff).
>
> Current issues are:
>
>    - "Jaggies" when looking up or down near a wall. Something with new
>    slopes not being consistent
>    - Slow
>    - When looking up, all terrain with floors is obscured. You couldn't
>    see a hulk looking at you from the roof.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/13595#issuecomment-170225927>
> .
>

kevingranade commented:
Segfault, go to an office tower in a release build, turn on debug mode, and try to look at z level 2.
Can't get anything out of the coredump, rebuilding with debug symbols.

Rivet-the-Zombie commented:
Always looking forward to more Z-level implementation.

Very nice, Coolthulhu!

kevingranade commented:
Segfault isn't a problem with the new code, drawing a level the player isn't on breaks an assumption of some new code.  (cata_tiles.cpp:1314)
It's trying to examine some data from a vehicle that's not on the current level.

kevingranade commented:
I'm hacking away at this, it ended up rather involved because there are a number of method calls that need a tripoint instead of a pair of x/y coordinates.

chaosvolt commented:
I'm not sure if that statement is implying a cause or a hindrance towards finding said cause.

chaosvolt commented:
Ah, nice to see that in. If I recall, this was the main thing keeping reality-devouring darkness ( https://github.com/CleverRaven/Cataclysm-DDA/issues/13786 ) in the "long term" pile.

stk2008 commented:
Ah was this PR to allow us to say..stand ontop of a tall building and see
the road etc below?

If so it dont work...well in TILES version :(

On Tue, Jan 12, 2016 at 8:37 AM, Chaosvolt <notifications@github.com> wrote:

> Ah, nice to see that in. If I recall, this was the main thing keeping
> reality-devouring darkness ( #13786
> <https://github.com/CleverRaven/Cataclysm-DDA/issues/13786> ) in the
> "long term" pile.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/13595#issuecomment-170838313>
> .
>

Malkeus commented:
It lets you use 'x' to look around. you use "<" and ">" to look up and down. The walking around vision is coming, if I'm reading the op correctly.

stk2008 commented:
oh damn I got to test this out then.

thanks

On Tue, Jan 12, 2016 at 3:38 PM, Malkeus <notifications@github.com> wrote:

> It lets you use 'x' to look around. you use "<" and ">" to look up and
> down. The walking around vision is coming, if I'm reading the op correctly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/13595#issuecomment-170949850>
> .
>

stk2008 commented:
Hmmm i press x the <> but it dont work?


On Tue, Jan 12, 2016 at 3:39 PM, daniel roberts <stkstkstk2008@gmail.com>
wrote:

> oh damn I got to test this out then.
>
> thanks
>
> On Tue, Jan 12, 2016 at 3:38 PM, Malkeus <notifications@github.com> wrote:
>
>> It lets you use 'x' to look around. you use "<" and ">" to look up and
>> down. The walking around vision is coming, if I'm reading the op correctly.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/CleverRaven/Cataclysm-DDA/pull/13595#issuecomment-170949850>
>> .
>>
>
>

Malkeus commented:
did you enable 3d vision in the debug menu?

Coolthulhu commented:
It is disabled by default, even on z-level worlds. You need to explicitly enable it in options.
That's because a lot of people are already playing with z-levels mode on and 3D vision is unfinished, very slow and might get buggy in some places.

Malkeus commented:
I'm excited about this one. Does enabling this mean zombies will be able to see me and path to me across z-levels now?

stk2008 commented:
i got it to work i think

You need to enable it in optoins and then enable DEBUG mode in game then I
can look around Z levels :)



On Tue, Jan 12, 2016 at 3:50 PM, Malkeus <notifications@github.com> wrote:

> I'm excited about this one. Does enabling this mean zombies will be able
> to see me and path to me across z-levels now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/13595#issuecomment-170953582>
> .
>

Coolthulhu commented:
See yes, but pathing may often be quite wonky.
